_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,607 @@
+```
+    @AfterAll
+    void teardownEnvironmentClass() {
+        if (Environment.SKIP_TEARDOWN == null) {
+            tearDownEnvironmentAfterAll();
+            teardownEnvForOperator();
+        }
+    }
+```
+
+so if you want to change teardown from your `@AfterAll`, you must override method `tearDownEnvironmentAfterAll()`:
+```
+    @Override
+    protected void tearDownEnvironmentAfterAll() {
+        doSomethingYouNeed();
+        super.tearDownEnvironmentAfterAll();
+    }
+```
+
+In order to delete all resources from specific `Resources` instance, execute:
+```
+    ResourceManager.deleteMethodResources();
+    ResourceManager.deleteClassResources();
+```
+
+## Auxiliary classes
+
+In our system test module we have many classes, which providing help. Each of them has different purpose and here is the list:
+
+1. Utils
+2. Constants
+3. Environment
+4. Kubernetes client
+5. Kafka clients
+6. Deployment file
+
+#### Utils
+
+These classes are applicable on many problems. We have separate utils for each resource. For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
+Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more. The hierarchy is as follows:
+
+* utils
+  * kafkaUtils
+    * KafkaBridgeUtils
+    * KafkaConnectorUtils
+    * KafkaConnectS2IUtils
+    * KafkaConnectUtils
+    * KafkaMirrorMaker2Utils
+    * KafkaMirrorMakerUtils
+    * KafkaRebalanceUtils
+    * KafkaTopicUtils
+    * KafkaUserUtils
+    * KafkaUtils
+  * kubeUtils
+    * controllers
+      * ConfigMapUtils
+      * DeploymentConfigUtils
+      * DeploymentUtils
+      * ReplicaSetUtils
+      * StatefulSetUtils
+    * objects
+      * NamespaceUtils
+      * PersistentVolumeClaimUtils
+      * PodUtils
+      * SecretUtils
+      * ServiceUtils
+  * specific
+    * BridgeUtils
+    * CruiseControlUtils
+    * KeycloakUtils
+    * MetricUtils
+    * TracingUtils
+    
+The usability of these classes is for instance when you have to to wait for some change, calling API, creating, changing, deleting Kubernetes resource and much more.
+The one method which deserves to be described is `waitFor`, where you will find it in many utils methods.
+You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
+that this value was updated but here is very good example where comes `race condition` into play. This race condition
+can occur if the assertion (verify phase) is too quick and value has not been updated. That's why we have these dynamic
+wait. The logic behind this method is that every `pollIntervalMs` for specific period of time `timeoutMs` the method checks
+`BooleanSupplier` condition which user can provide and if this condition is true the method will successfully end. Otherwise, we will get the
+`WaitException` with some reasonable description. In the Figure 4.1 we can see the whole `waitFor` method.
```

Comment:
I don't know how much is really useful posting the whole code for the method. Just describing its purpose, the parameters meaning should be ok. If the reader wants, he can just go to the source code in the repo (maybe you can add the link here).

Reply:
Make sense :)